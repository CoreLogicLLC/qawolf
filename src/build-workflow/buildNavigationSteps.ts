import { Action, WindowEvent, Step } from '../types';

export const buildNavigationSteps = (events: WindowEvent[]): Step[] => {
  return events.map((event, index) => ({
    action: (event.name as Action),
    event,
    index,
    value: event.value,
  }));
};
