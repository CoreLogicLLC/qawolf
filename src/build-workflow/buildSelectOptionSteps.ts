import { ElementEvent, InputEvent, Step } from '../types';

export const buildSelectOptionSteps = (events: ElementEvent[]): Step[] => {
  const steps: Step[] = [];

  for (let i = 0; i < events.length; i++) {
    const event = events[i] as InputEvent;

    // ignore system initiated actions & other non-input actions
    if (!event.isTrusted || event.name !== 'input') continue;

    // ignore input events not on selects
    if (event.target.name !== 'select') continue;

    steps.push({
      action: 'selectOption',
      event,
      index: steps.length,
      value: event.value,
    });
  }

  return steps;
};
