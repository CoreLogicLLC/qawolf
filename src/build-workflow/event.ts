import { ElementEvent } from '../types';

export const isChangeEvent = (event: ElementEvent | null): boolean =>
  event && event.isTrusted && event.name === 'change';

export const isInputEvent = (event: ElementEvent | null): boolean =>
  event && event.isTrusted && event.name === 'input';

export const isKeyEvent = (event: ElementEvent | null): boolean =>
  event &&
  event.isTrusted &&
  (event.name === 'keydown' || event.name === 'keyup');

export const isPasteEvent = (event: ElementEvent | null): boolean =>
  event && event.isTrusted && event.name === 'paste';
