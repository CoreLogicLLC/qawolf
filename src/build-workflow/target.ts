import { Doc } from '../types';

export const isContentEditableTarget = (target: Doc): boolean => {
  const { contenteditable } = target.attrs || {};
  return contenteditable && contenteditable !== "false";
}

export const isInputTarget = (target: Doc): boolean => {
  const name = target.name || '';
  return name.toLowerCase() === 'input';
}

export const isTextareaTarget = (target: Doc): boolean => {
  const name = target.name || '';
  return name.toLowerCase() === 'textarea';
}
