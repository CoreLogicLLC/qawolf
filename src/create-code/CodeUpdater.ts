import Debug from 'debug';
import { EventEmitter } from 'events';
import { buildVirtualCode } from '../build-code/buildVirtualCode';
import { CodeReconciler } from './CodeReconciler';
import { getLineIncludes, removeLinesIncluding } from './format';
import { CREATE_HANDLE, PATCH_HANDLE } from './patchCode';
import { Step } from '../types';

const debug = Debug('qawolf:CodeUpdater');

export type CodeFileOptions = {
  path: string;
};

type UpdateOptions = {
  steps: Step[];
};

export abstract class CodeUpdater extends EventEmitter {
  protected _locked = false;
  private _reconciler: CodeReconciler;

  protected constructor() {
    super();
    this._reconciler = new CodeReconciler();
  }

  protected abstract _loadCode(): Promise<string>;
  protected abstract _updateCode(code: string): Promise<void>;

  private async _update(code: string): Promise<void> {
    debug('update code');
    this.emit('codeupdate', code);
    await this._updateCode(code);
  }

  protected async _prepare(): Promise<void> {
    const code = await this._loadCode();

    const createLine = getLineIncludes(code, CREATE_HANDLE);
    if (!createLine) return;

    const updatedCode = code.replace(createLine.trim(), PATCH_HANDLE);
    await this._update(updatedCode);
  }

  public async finalize(): Promise<void> {
    this._locked = true;
    let code = await this._loadCode();
    code = removeLinesIncluding(code, PATCH_HANDLE);
    await this._update(code);
  }

  public async update(options: UpdateOptions): Promise<void> {
    // do not conflict with an update in progress
    if (this._locked) {
      debug(`skip update: update in progress`);
      return;
    }

    // check the virtual code changed
    const updatedVirtualCode = buildVirtualCode(options.steps);
    if (!this._reconciler.hasChanges(updatedVirtualCode)) {
      debug(`skip update: no virtual changes`);
      return;
    }

    this._locked = true;

    // update the actual code
    const actualCode = await this._loadCode();

    const updatedCode = this._reconciler.reconcile({
      actualCode,
      virtualCode: updatedVirtualCode,
    });
    await this._update(updatedCode);

    // store the updated virtual code
    this._reconciler.update(updatedVirtualCode);

    this._locked = false;
  }
}
