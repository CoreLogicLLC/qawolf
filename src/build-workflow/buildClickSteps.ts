import { isInputTarget } from './target';
import { ElementEvent, Step } from '../types';

const CLICK_EVENTS = ['click', 'mousedown'];

const shouldIncludeClickEvent = (
  event: ElementEvent,
  previousEvent?: ElementEvent,
): boolean => {
  // ignore system initiated clicks
  if (!event.isTrusted) return false;

  // ignore non-click events
  if (!CLICK_EVENTS.includes(event.name)) return false;

  // ignore clicks on selects
  if (event.target.name === 'select') return false;

  // skip system-initiated clicks triggered by a key press
  // ex. "Enter" triggers a click on a submit input
  if (
    previousEvent &&
    ['change', 'keydown', 'keyup'].includes(previousEvent.name) &&
    event.time - previousEvent.time < 50
  ) {
    return false;
  }

  return true;
};

const groupClickEvents = (events: ElementEvent[]): ElementEvent[][] => {
  const clickGroups = [];

  let group: ElementEvent[] = [];

  events.forEach((event, i) => {
    const previousEvent = events[i - 1];

    if (!shouldIncludeClickEvent(event, previousEvent)) return;

    // always group clicks with the previous mousedown/click events
    // a click will follow a mousedown if the mouse was released within the same element
    // a click will follow a click if it propagated to a higher element (a click on a label propagates to the input)
    if (!group.length || event.name === 'click') {
      group.push(event);
      return;
    }

    // start a new group
    clickGroups.push(group);
    group = [event];
  });

  if (group.length) {
    // append last group
    clickGroups.push(group);
  }

  return clickGroups;
};

export const buildClickSteps = (events: ElementEvent[]): Step[] => {
  const groupedClickEvents = groupClickEvents(events);
  const steps: Step[] = [];

  groupedClickEvents.forEach((events) => {
    let event = events[0] as ElementEvent;

    const inputEvent = events.find((event) => isInputTarget(event.target));
    if (inputEvent) {
      // if an event in the group is on an input, assume the click propagated
      // to an element like a checkbox or radio, which is most accurate target
      event = inputEvent;
    }

    steps.push({
      action: 'click',
      event,
      index: steps.length,
    });
  });

  return steps;
};
