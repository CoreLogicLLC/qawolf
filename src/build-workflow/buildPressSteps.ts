import Debug from 'debug';
import {
  Doc,
  ElementEvent,
  KeyEvent,
  Step,
} from '../types';
import { isContentEditableTarget, isInputTarget, isTextareaTarget } from './target';

const debug = Debug('qawolf:buildPressSteps');

/**
 * The full list:
 * https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
 */

const KEYS_TO_TRACK_FOR_CONTENTEDITABLE = new Set([
  // Enter types a line break, shouldn't be a press.
  'Escape',
  // NOTE: Sometimes tab types a tab character, but this seems to be
  // only with libraries like Quill that must intercept the keydown.
  // It seems difficult to detect when this is happening, so for now
  // an extra Tab press may be built and would need to be manually
  // deleted from the generated test.
  'Tab'
]);

const KEYS_TO_TRACK_FOR_INPUT = new Set([
  'Enter',
  'Escape',
  'Tab'
]);

const KEYS_TO_TRACK_FOR_TEXTAREA = new Set([
  // Enter types a line break, shouldn't be a press.
  'Escape',
  'Tab'
]);

/**
 * These are key presses that we want to include when playing back as long as they
 * aren't being pressed as part of editing some input text.
 */
const KEYS_TO_TRACK_FOR_NON_INPUT = new Set([
  'ArrowDown',
  'ArrowLeft',
  'ArrowRight',
  'ArrowUp',
  'Backspace',
  'Delete',
  'End',
  'Enter',
  'Escape',
  'Home',
  'PageDown',
  'PageUp',
  'Tab'
]);

/**
 * @summary Determines whether a key press event should be included as part of the
 *   test code being built. Some keys are always included and some are included only
 *   when we're not editing text, i.e. the target isn't an input of some sort.
 * @param {String} key The name of the key that was pressed.
 * @param {Object} target Some details about the event target.
 * @return {Boolean} True if we should include this key as part of the playback script.
 */
const shouldTrackKeyPress = (key: string, target: Doc): boolean => {
  if (isInputTarget(target)) return KEYS_TO_TRACK_FOR_INPUT.has(key);
  if (isTextareaTarget(target)) return KEYS_TO_TRACK_FOR_TEXTAREA.has(key);
  if (isContentEditableTarget(target)) return KEYS_TO_TRACK_FOR_CONTENTEDITABLE.has(key);
  return KEYS_TO_TRACK_FOR_NON_INPUT.has(key);
}

/**
 * @summary Given a list of captured browser page events, returns a list of key press
 *   steps that should be included when playing back the flow for testing purposes.
 */
export const buildPressSteps = (events: ElementEvent[]): Step[] => {
  debug('building press steps');

  return events.reduce((pressSteps: Step[], event: ElementEvent): Step[] => {
    if (event.name === 'keydown') {
      const { value } = event as KeyEvent;
      if (shouldTrackKeyPress(value, event.target)) {
        debug(`build press step for keydown ${value}`);
        pressSteps.push({
          action: 'press',
          event,
          index: pressSteps.length,
          value,
        });
      } else {
        debug(`not building press step for keydown ${value}`);
      }
    }

    return pressSteps;
  }, []);
};
