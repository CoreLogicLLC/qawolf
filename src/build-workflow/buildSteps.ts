import { concat, sortBy } from 'lodash';
import { buildClickSteps } from './buildClickSteps';
import { buildFillSteps } from './buildFillSteps';
import { buildNavigationSteps } from './buildNavigationSteps';
import { buildPressSteps } from './buildPressSteps';
import { buildScrollSteps } from './buildScrollSteps';
import { buildSelectOptionSteps } from './buildSelectOptionSteps';
import { ElementEvent, Step, WindowEvent } from '../types';

export const buildSteps = (elementEvents: ElementEvent[], windowEvents: WindowEvent[] = []): Step[] => {
  const unorderedSteps = concat(
    buildClickSteps(elementEvents),
    buildFillSteps(elementEvents),
    buildPressSteps(elementEvents),
    buildScrollSteps(elementEvents),
    buildSelectOptionSteps(elementEvents),
    buildNavigationSteps(windowEvents),
  );

  let steps = sortBy(
    unorderedSteps,
    // ordered by the event time
    (step) => step.event.time,
  );

  // reindex
  steps = steps.map((step, index) => ({
    ...step,
    index,
  }));

  return steps;
};
