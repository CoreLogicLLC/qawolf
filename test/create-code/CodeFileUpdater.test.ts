import { outputFile, pathExists, readFile, remove } from 'fs-extra';
import { buildSteps } from '../../src/build-workflow/buildSteps';
import { CodeFileUpdater } from '../../src/create-code/CodeFileUpdater';
import { CREATE_HANDLE, PATCH_HANDLE } from '../../src/create-code/patchCode';

// require manually since fs is mocked
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { events } = require('../fixtures/login.json');
const steps = buildSteps(events);

jest.mock('fs-extra');

(pathExists as jest.Mock).mockResolvedValue(true);

const mockedReadFile = readFile as jest.Mock;
const mockedRemove = remove as jest.Mock;
const mockedOutputFile = outputFile as jest.Mock;

describe('CodeFileUpdater', () => {
  beforeEach(() => {
    mockedOutputFile.mockReset();
  });

  describe('discard', () => {
    it('reverts to initial code', async () => {
      const initialCode = `initialCode()\n${CREATE_HANDLE}`;
      mockedReadFile.mockResolvedValue(initialCode);
      const updater = await CodeFileUpdater.create('revertpath');
      await updater.discard();

      const reverted = mockedOutputFile.mock.calls[1];
      expect(reverted[0]).toEqual('revertpath');
      expect(reverted[1]).toEqual(initialCode);
      expect(mockedRemove.mock.calls.length).toEqual(0);
    });

    it('removes code if QAW_CREATE is set to true', async () => {
      process.env.QAW_CREATE = 'true';
      const initialCode = `initialCode()\n${CREATE_HANDLE}`;
      mockedReadFile.mockResolvedValue(initialCode);
      const updater = await CodeFileUpdater.create('removepath');
      await updater.discard();
      expect(mockedRemove.mock.calls[0][0]).toEqual('removepath');
      process.env.QAW_CREATE = undefined;
    });
  });

  describe('create', () => {
    it('replaces qawolf.create call with the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `  someCode();\n  await qawolf.create();\n  otherCode();`,
      );
      await CodeFileUpdater.create('somepath');

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe(
        `  someCode();\n  ${PATCH_HANDLE}\n  otherCode();`,
      );
    });
  });

  describe('finalize', () => {
    it('removes the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `someCode();\n${PATCH_HANDLE}\notherCode();`,
      );

      const updater = await CodeFileUpdater.create('somepath');
      await updater.finalize();

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe('someCode();\notherCode();');
    });
  });

  describe('update', () => {
    it('saves code with new steps', async () => {
      mockedReadFile.mockResolvedValue(
        `  someCode();\n  ${PATCH_HANDLE}\n  otherCode();`,
      );

      const updater = await CodeFileUpdater.create('somepath');

      const codeUpdateEvents = [];

      updater.on('codeupdate', (code) => {
        codeUpdateEvents.push(code);
      });

      // emulate only the first two steps being ready
      await updater.update({ steps: steps.slice(0, 2) });

      const codeUpdateOne = mockedOutputFile.mock.calls[0][1];
      expect(codeUpdateOne).toMatchSnapshot();
      expect(codeUpdateEvents[0]).toMatch(codeUpdateOne);
      mockedReadFile.mockResolvedValueOnce(codeUpdateOne);

      await updater.update({ steps });
      const codeUpdateTwo = mockedOutputFile.mock.calls[1][1];
      expect(codeUpdateEvents[1]).toMatch(codeUpdateTwo);
      expect(codeUpdateTwo).toMatchSnapshot();
    });
  });
});
