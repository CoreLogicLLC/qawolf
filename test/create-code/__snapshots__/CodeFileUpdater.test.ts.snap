// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeFileUpdater update saves code with new steps 1`] = `
"  someCode();
  const page = await qawolf.waitForPage(context, 0, { waitUntil: \\"domcontentloaded\\" });
  await page.click(\\"text=Log in\\");
  await page.click(\\"#username\\");
  // üê∫ create code here
  otherCode();"
`;

exports[`CodeFileUpdater update saves code with new steps 2`] = `
"  someCode();
  const page = await qawolf.waitForPage(context, 0, { waitUntil: \\"domcontentloaded\\" });
  await page.click(\\"text=Log in\\");
  await page.click(\\"#username\\");
  await page.fill(\\"#username\\", \\"tomsmith\\");
  await page.press(\\"#username\\", \\"Tab\\");
  await page.fill(\\"#password\\", \\"SuperSecretPassword!\\");
  await page.press(\\"#password\\", \\"Enter\\");
  await page.click(\\"text=Log out\\");
  // üê∫ create code here
  otherCode();"
`;
